spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/SimpleMessageListenerContainer.java

```diff
@@ -770,13 +777,35 @@ private boolean doReceiveAndExecute(BlockingQueueConsumer consumer) throws Throw
 				executeListener(channel, message);
 			}
 			catch (ImmediateAcknowledgeAmqpException e) {
+				if (this.logger.isDebugEnabled()) {
+					this.logger.debug("User requested ack for failed delivery");
+				}
 				break;
 			}
 			catch (Throwable ex) { //NOSONAR
-				consumer.rollbackOnExceptionIfNecessary(ex);
-				throw ex;
+				if (causeChainHasImmediateAcknowledgeAmqpException(ex)) {
+					if (this.logger.isDebugEnabled()) {
+						this.logger.debug("User requested ack for failed delivery");
+					}
+					break;
+				}
+				if (getTransactionManager() != null) {
+					if (getTransactionAttribute().rollbackOn(ex)) {
+						consumer.rollbackOnExceptionIfNecessary(ex);
+						throw ex;
+					}
+					else {
+						if (this.logger.isDebugEnabled()) {
+							this.logger.debug("No rollback for " + ex);
+						}
+						break;
```

Comment:
Here we're running on the async consumer thread so we need to throw. See my comment for the DMLC case.